### PR TITLE
fix: reject null hash values on delete + primary-key changes on populated tables (studio#1199)

### PR DESCRIPTION
**Pull request by dawsontoth**
Two related data-integrity fixes behind **HarperFast/studio#1199** ("500 'hash_values' must be strings or numbers"). Studio companion PR: **HarperFast/studio#1529**.

---

## 1. `delete` with a null hash value wipes the whole table

A `delete` with a null/undefined value in `hash_values` deletes **every record in the table** while reporting `"1 of 1 record successfully deleted"`:

```jsonc
{"operation":"delete","database":"data","table":"T","hash_values":[null]}
// → {"message":"1 of 1 record successfully deleted","deleted_hashes":[null]}  (200)
// → table now has 0 rows
```

Reproduced on 4.7.23 and 5.0.15. `hash_values:[]` is a safe no-op and `[999]` (absent key) deletes nothing — only `null` triggers the wipe. A null id is coerced by `transactional` into a whole-collection target (only `get` was guarded against null), so `Table.delete` takes the search-and-delete-all branch and `search()` runs with no key bound.

**Fix:** reject `null`/`undefined` items in `deleteValidator` (composite array/object keys still allowed), and guard each id in `ResourceBridge.deleteRecords` before `Table.delete` so no caller can trip the coercion. Tests at both layers.

## 2. Changing `@primaryKey` on a populated table corrupts it

Changing which attribute is the primary key (via GraphQL schema, `defineTable`, or `create_table`) on a table that already has records leaves the table in a lying state: the real hash/storage key never moves, so `describe_table` reports the newly-declared attribute while **every** record — old *and* newly inserted, even through the raw Operations API — stays keyed by the original attribute. `search_by_id`/update/delete by the declared key then all miss. I confirmed this by inserting `{email:"…"}` into a table whose `describe` said `primary_key: "email"` and finding the row keyed by a generated `id` UUID.

This is the actual origin of #1199 (a row's declared primary key resolves to nothing → studio builds doomed requests).

**Fix:** `table()` is the single factory every front-end funnels through. On the existing-table branch, if a schema-authored declaration names a different primary key **and** the table has any records, throw a `ClientError` instead of silently diverging describe from storage. Empty tables and settings-only reasserts are unaffected. Regression test in `unitTests/resources/update-schema.test.js`.

> Scope note: this rejects the corrupting change with a clear error. Actually *migrating* (re-keying existing records to the new attribute) is a larger feature and intentionally out of scope here.

---

Existing delete/schema tests remain green.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: dataLayer/harperBridge/ResourceBridge.ts [hunk 1]
```diff
@@ -282,6 +282,13 @@ export class ResourceBridge extends BridgeMethods {
 		return transaction(context, async (transaction) => {
 			const ids: Id[] =
 				deleteObj.ids || deleteObj.hash_values || deleteObj.records.map((record) => record[Table.primaryKey]);
+			// A null/undefined hash value must never reach Table.delete: with no key bound it is coerced
+			// into a whole-collection target and silently deletes EVERY record in the table (reported as
+			// "1 of 1 record successfully deleted"). Reject before touching anything, mirroring the
+			// primary-key guard search_by_id already enforces. See HarperFast/studio#1199.
+			for (const id of ids) {
+				if (id == null) throw new ClientError('Invalid primary key of null', 400);
```
Comment (cb1kenobi) [inline dataLayer/harperBridge/ResourceBridge.ts:290]: **Low: layer-2 guard has no test — the SQL/internal delete path is uncovered**

This guard is the *only* protection for callers that skip `deleteValidator`: the SQL `DELETE` path (`sqlTranslator/deleteTranslator.ts` calls `harperBridge.deleteRecords` directly) and any internal caller. Yet both new tests exercise only the validator layer — `delete.test.js` stubs `harperBridge` entirely, so this line never runs under test. If a refactor drops it, the SQL delete-by-null-PK wipe (the studio#1199 scenario, reachable when a record's `@primaryKey` attribute is unset) silently regresses with no failing test.

Suggested fix: add a `deleteRecords` test that passes an id array containing `null` (or a `records` entry whose PK maps to `null`) and asserts it throws `Invalid primary key of null` before any `Table.delete` call.

—
_Generated by Barber AI_

File: validation/deleteValidator.ts [hunk 2]
```diff
@@ -2,12 +2,19 @@ import * as validator from './validationWrapper.ts';
 import Joi from 'joi';
 import { hdbTable, hdbDatabase } from './common_validators.ts';
 
+// `hash_values`/`ids` are lists of primary keys. Items may be strings, numbers, or composite
+// (array/object) keys, but never `null`/`undefined`: a null key is coerced downstream into a
+// whole-collection target that wipes the table (see HarperFast/studio#1199), reject at the boundary.
+const hashValue = Joi.any().invalid(null).messages({
+	'any.invalid': `'hash_values' must not contain null`,
+});
```
[the remark this answers is not in the document]
Comment (dawsontoth) [reply validation/deleteValidator.ts:12]: Done in e92e5378 — the message now uses `{#label}`, so a null in `ids` reports `'ids[0]' must not contain null` (and `hash_values` reports `'hash_values[N]' ...`). Kept `{#label}` rather than a bare string, since Joi doesn't auto-prepend the field name otherwise (verified: a bare `'must not contain null'` produces no field name at all).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: unitTests/validation/deleteValidator.test.js [hunk 3]
```diff
@@ -32,4 +32,35 @@ describe('Test deleteValidator module', () => {
...
+	it('Test null in ids rejected', () => {
+		const test_del_obj = {
+			schema: 'unit',
+			table: 'test',
+			ids: [null],
+			hash_values: [1],
+		};
+		const result = deleteValidator(test_del_obj);
+		expect(result.message).to.equal("'hash_values' must not contain null");
+	});
```
[the remark this answers is not in the document]
Comment (dawsontoth) [reply unitTests/validation/deleteValidator.test.js]: Done in e92e5378 — converted the file to bare `node:assert` and updated the assertions to the new messages. Used bare `node:assert` rather than `node:assert/strict`: AGENTS.md mandates it and oxlint's `no-restricted-imports` actually rejects `assert/strict` here (`assert.strictEqual` still works on the bare module).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: dataLayer/harperBridge/ResourceBridge.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dawsontoth) [reply dataLayer/harperBridge/ResourceBridge.ts:290]: Done in e92e5378 — added `unitTests/resources/deleteNullPrimaryKey.test.js` that drives the real `ResourceBridge.deleteRecords` (no stubbing) with a null id via `hash_values`, `ids`, AND a `records` entry missing its primary key (the SQL DELETE shape that bypasses `deleteValidator`), asserting it throws `Invalid primary key of null` and deletes nothing — plus a valid-id case so the guard isn't over-blocking.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

Comment (dawsontoth) [reply dataLayer/harperBridge/ResourceBridge.ts:290]: Done in e92e5378 — `unitTests/resources/deleteNullPrimaryKey.test.js` now exercises this guard directly against the real `ResourceBridge.deleteRecords` (no `harperBridge` stub): null id via `hash_values`, `ids`, and a `records` entry whose PK maps to null (the SQL delete path), each asserting it throws before any `Table.delete` and the table still has its rows.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by kriszyp** [comment]
Pushed one more commit (85316bf) extending this fix a layer up: the Resource API had the same footgun — `Table.delete(null)` (or `undefined`, or no argument) was normalized by `transactional()` into a whole-collection target with no conditions and silently deleted every record, returning `true`. The ops-API/bridge guards in this PR didn't cover that path, since component code calls the Resource API directly.

The commit mirrors the existing `get` guard in `Resource.ts` so a bare `null`/`undefined` id on `delete` throws a 400 `ClientError`, and adds `unitTests/resources/deleteNullResourceApi.test.js` covering the three rejected forms plus the still-working delete-by-id and deliberate delete-all via an explicit query object. Great catch on the original bug — the validator + bridge guards here were exactly the right layers for the ops/SQL paths, and this just closes the last entry point for the same bug class.

— Claude (KrAIs)

**Review by kriszyp — APPROVED** [review]
I went a little deeper with preventing `Table.delete(null)`, adding a commit, I hope that is ok.
Changing the primary key of a table, even if it is empty, sounds a bit spicy, but if it works, that's great.
Looks good!

**Review by cb1kenobi — APPROVED** [review]
But what if I want to wipe the whole table? Seems so much easier than `{ "operation": "sql",  "sql": "DELETE FROM your_database.your_table" }`.

**Comment by kriszyp** [comment]
> But what if I want to wipe the whole table? Seems so much easier than `{ "operation": "sql", "sql": "DELETE FROM your_database.your_table" }`.

You should still be able to delete all records (easily?) with `Table.delete({})`.

**Comment by dawsontoth** [comment]
You can also drop the table and restart your cluster, it'll pop back into existence. This is the operation the UI uses to wipe out a table.
```
{
  operation: 'drop_table',
  database,
  table,
  replicated,
}
```
